File: Python/textbook_track/chapter05/ba5d.py
# Find the Longest Path in a DAG
import functools as ft
import sys

from collections import defaultdict
from itertools import count
from math import inf


class DirectedWeightedGraph:
    def __init__(self, adjacency_list):
        self._adjacency_list = adjacency_list
        self._nodes = self._collect_graph_nodes()
        self._incoming_edges = self._calc_incoming_edges()

        self._previsit_numbers = {}
        self._previsit_id = count(start=1)

    @property
    def incoming_edges(self):
        return self._incoming_edges

    @ft.cached_property
    def topological_ordering(self):
        return self._run_depth_first_search()

    def _collect_graph_nodes(self):
        nodes = set()
        for node, neighbours in self._adjacency_list.items():
            nodes.add(node)
            for neighbour, _ in neighbours:
                nodes.add(neighbour)
        return frozenset(nodes)

    def _calc_incoming_edges(self):
        ingraph = defaultdict(list)
        for node, neighbours in self._adjacency_list.items():
            for neighbour, weight in neighbours:
                ingraph[neighbour].append((node, weight))
        return dict(ingraph)

    def _run_depth_first_search(self):
        topological_sorting = []
        for node in self._nodes:
            if not self._is_visited_in_dfs(node):
                self._explore(node, topological_sorting)
        return topological_sorting[::-1]

    def _is_visited_in_dfs(self, node):
        return self._previsit_numbers.get(node, 0) > 0

    def _find_unvisited_neighbour_of_a_node(self, node):
        neighbours = self._adjacency_list.get(node, [])
        for neighbour, _ in neighbours:
            if not self._is_visited_in_dfs(neighbour):
                return neighbour
        return None

    def _explore(self, starting_node, topological_sorting):
        self._previsit_numbers[starting_node] = next(self._previsit_id)
        previsit_stack = [starting_node]
        while previsit_stack:
            last_node = previsit_stack.pop(-1)
            unvisited_neighbour = self._find_unvisited_neighbour_of_a_node(last_node)
            if unvisited_neighbour is None:
                topological_sorting.append(last_node)
            else:
                self._previsit_numbers[unvisited_neighbour] = next(self._previsit_id)
                previsit_stack.append(last_node)
                previsit_stack.append(unvisited_neighbour)


def read_weighted_edges(lines):
    adjacency_list = defaultdict(list)
    for line in lines:
        node_from, node_to, weight = map(int, line.replace('->', ':').split(':'))
        adjacency_list[node_from].append((node_to, weight))
    return dict(adjacency_list)


def collect_longest_path(backtrack, source, sink):
    path = [sink]
    node = sink
    while node != source:
        node = backtrack[node]
        path.append(node)
    return path[::-1]


def find_longest_path(graph, source, sink):
    longest_path = {}
    backtrack = {}
    for node in graph.topological_ordering:
        predecessors = graph.incoming_edges.get(node, [])
        if node == source:
            longest_path[node] = 0
        elif predecessors:
            paths_from_predecessors = ((u, longest_path.get(u, -inf) + w) for u, w in predecessors)
            v, s = max(paths_from_predecessors, key=lambda x: x[1])
            longest_path[node] = s
            backtrack[node] = v

    return longest_path[sink], collect_longest_path(backtrack, source, sink)


def main():
    data = sys.stdin.read().splitlines()
    source = int(data[0])
    sink = int(data[1])
    adjacency_list = read_weighted_edges(data[2:])
    graph = DirectedWeightedGraph(adjacency_list)
    length, longest_path = find_longest_path(graph, source, sink)
    print(length)
    print('->'.join(map(str, longest_path)))


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter05/BA5D.scala
package TextbookTrack.Chapter05

object BA5D {
  import scala.annotation.tailrec
  import scala.collection.mutable.{Map => MutableMap}

  final case class WeightedEdge(nodeFrom: Int, nodeTo: Int, weight: Int)

  class DirectedWeightedGraph(val adjacencyList: Map[Int, List[WeightedEdge]]) {
    import DirectedWeightedGraph.depthFirstSearch

    private val nodes: Set[Int] = adjacencyList.keySet ++ adjacencyList.valuesIterator.flatMap(_.map(_.nodeTo)).toSet

    val incomingEdges: Map[Int, List[WeightedEdge]] =
      (for {
        (node, neighbours) <- adjacencyList.iterator
        WeightedEdge(_, nodeTo, weight) <- neighbours
      } yield WeightedEdge(nodeTo, node, weight)).toList.groupBy(_.nodeFrom)

    lazy val topologicalSorting: List[Int] = depthFirstSearch(this)
  }

  object DirectedWeightedGraph {
    private def depthFirstSearch(graph: DirectedWeightedGraph): List[Int] = {
      val visitStarted: MutableMap[Int, Int] = MutableMap()
      val previsitId: Iterator[Int] = Iterator.from(1)

      def isNodeVisited(node: Int): Boolean = visitStarted.getOrElse(node, 0) > 0

      def findUnvisitedNeighbour(node: Int): Option[Int] =
        graph.adjacencyList.getOrElse(node, Nil).map(_.nodeTo).find(!isNodeVisited(_))

      def explore(startingNode: Int, topologicalSorting: List[Int]): List[Int] = {
        @tailrec
        def loop(acc: List[Int], previsitStack: List[Int]): List[Int] = previsitStack match {
          case Nil => acc
          case lastNode :: rest =>
            findUnvisitedNeighbour(lastNode) match {
              case None => loop(lastNode :: acc, rest)
              case Some(unvistedNeighbour) =>
                visitStarted(unvistedNeighbour) = previsitId.next()
                loop(acc, unvistedNeighbour :: previsitStack)
            }
        }

        visitStarted(startingNode) = previsitId.next()
        loop(topologicalSorting, List(startingNode))
      }

      @tailrec
      def run(topologicalSorting: List[Int], nodes: Set[Int]): List[Int] = {
        nodes.headOption match {
          case None => topologicalSorting
          case Some(node) =>
            val updatedSorting: List[Int] =
              if (isNodeVisited(node)) topologicalSorting else explore(node, topologicalSorting)
            run(updatedSorting, nodes.tail)
        }
      }
      run(Nil, graph.nodes)
    }
  }

  private def splitEdge(line: String): WeightedEdge = line.replace("->", ":").split(":").map(_.toInt).toList match {
    case List(from, to, weight) => WeightedEdge(from, to, weight)
    case _ => throw new Exception("Malformed input data.")
  }

  private def collectLongestPath(backtrack: Map[Int, Int], source: Int, sink: Int): List[Int] = {
    @tailrec
    def loop(acc: List[Int], currentNode: Int): List[Int] = {
      if (currentNode == source) acc
      else {
        val nextNode: Int = backtrack(currentNode)
        loop(nextNode :: acc, nextNode)
      }
    }
    loop(List(sink), sink)
  }

  def readWeightedEdges(lines: Iterator[String]): Map[Int, List[WeightedEdge]] =
    lines.map(splitEdge).toList.groupBy(_.nodeFrom)

  def findLongestPath(graph: DirectedWeightedGraph, source: Int, sink: Int): (Int, List[Int]) = {
    @tailrec
    def loop(longestPath: Map[Int, Double],
             backtrack: Map[Int, Int],
             orderedNodes: List[Int]): (Map[Int, Double], Map[Int, Int]) = orderedNodes match {
      case Nil => (longestPath, backtrack)
      case node :: rest =>
        graph.incomingEdges.get(node) match {
          case Some(predecessors) if node != source =>
            val paths: List[(Int, Double)] = predecessors.map{
              case WeightedEdge(_, u, w) => (u, longestPath.getOrElse(u, Double.NegativeInfinity) + w)
            }
            val (v, s): (Int, Double) = paths.maxBy{ case (_, cost) => cost }
            loop(longestPath + (node -> s), backtrack + (node -> v), rest)
          case _ => loop(longestPath, backtrack, rest)
        }
    }
    val (longestPath, backtrack): (Map[Int, Double], Map[Int, Int]) =
      loop(Map(source -> 0), Map(), graph.topologicalSorting)
    (longestPath(sink).toInt, collectLongestPath(backtrack, source, sink))
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val source: Int = reader.next().toInt
    val sink: Int = reader.next().toInt
    val adjacencyList: Map[Int, List[WeightedEdge]] = readWeightedEdges(reader)
    val graph = new DirectedWeightedGraph(adjacencyList)
    val (length, longestPath): (Int, List[Int]) = findLongestPath(graph, source, sink)
    println(length)
    println(longestPath.mkString("->"))
  }
}
